### PR TITLE
Use rules condition containers.

### PR DESCRIPTION
**Pull request by fubhy**

### CODE REVIEW

File: src/Plugin/RulesExpression/Rule.php [hunk 1]
```diff
@@ -55,7 +93,10 @@ public function execute() {
    * {@inheritdoc}
    */
   public function addCondition(RulesConditionInterface $condition) {
-    $this->conditions[] = $condition;
+    if (!isset($this->conditions)) {
+      $this->conditions = $this->expressionManager->createInstance('rules_and');
```
Comment (klausi) [inline src/Plugin/RulesExpression/Rule.php]: This should have a comment like "Per default we use an AND condition container. Any condition added will be AND-joined to the others."

**Comment by fubhy** [comment]
re-rolled.

File: tests/src/RuleTest.php [hunk 2]
```diff
@@ -25,6 +40,18 @@ public static function getInfo() {
...
 
+  public function setUp() {
+    parent::setUp();
+
+    $this->expressionManager = $this->getMockBuilder('Drupal\rules\Plugin\RulesExpressionPluginManager')
+      ->disableOriginalConstructor()
+      ->getMock();
+
+    $this->testRule = new Rule([], 'test', [], $this->expressionManager);
+    $this->testRule->setConditions(new RulesAnd(array(), 'test', array()));
```
Comment (klausi) [inline tests/src/RuleTest.php]: array() should not be used, use [].

File: tests/src/RuleTest.php [hunk 3]
```diff
@@ -35,10 +62,9 @@ public function testActionExecution() {
 
     // Create a test rule, we don't care about plugin information in the
     // constructor.
```
Comment (klausi) [inline tests/src/RuleTest.php]: Comment does not fit to the code below.

File: tests/src/RuleTest.php [hunk 4]
```diff
@@ -51,10 +77,9 @@ public function testConditionFails() {
...
```
Comment (klausi) [inline tests/src/RuleTest.php]: Same here, outdated comment.

File: tests/src/RuleTest.php [hunk 5]
```diff
@@ -25,6 +40,18 @@ public static function getInfo() {
...
   }
 
+  public function setUp() {
+    parent::setUp();
+
+    $this->expressionManager = $this->getMockBuilder('Drupal\rules\Plugin\RulesExpressionPluginManager')
+      ->disableOriginalConstructor()
+      ->getMock();
+
+    $this->testRule = new Rule([], 'test', [], $this->expressionManager);
```
Comment (klausi) [inline tests/src/RuleTest.php]: this should have the comment that we don't care about plugin config here, hence the empty arrays and 'test'.

**Comment by fubhy** [comment]
This will need yet another re-roll once the minor adjustments from https://github.com/fago/rules/pull/34 and https://github.com/fago/rules/pull/33 are in.

File: src/Plugin/RulesExpression/Rule.php [hunk 6]
```diff
@@ -36,15 +40,42 @@ class Rule extends PluginBase implements RuleInterface {
...
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param array $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\rules\Plugin\RulesExpressionPluginManager $expression_manager
+   *   The rules expression plugin manager.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, RulesExpressionPluginManager $expression_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->conditions = $expression_manager->createInstance('rules_and');
```
Comment (klausi) [inline src/Plugin/RulesExpression/Rule.php:61]: This should have a comment "Per default the outer condition container of a rule is initialized as conjunction (AND), meaning that all conditions in it must evaluate to TRUE to fire the actions."

File: src/Plugin/RulesExpression/Rule.php [hunk 7]
```diff
@@ -36,15 +40,42 @@ class Rule extends PluginBase implements RuleInterface {
...
    */
   public function execute() {
     // Evaluate conditions, if they pass execute actions.
-    foreach ($this->conditions as $condition) {
-      if (!$condition->execute()) {
-        // If a condition returns FALSE stop here.
-        return;
-      }
+    if (!$this->conditions->execute()) {
+      // If a condition returns FALSE stop here.
```
Comment (klausi) [inline src/Plugin/RulesExpression/Rule.php]: Outdated comment, we are not iterating over conditions anymore. Should be "The conditions of this rule are not satisfied, therefore execution of the rule is stopped here immediately."

**Pull request merged**